File: frontend/lib/steward.ts
export type StewardPipelineStage = "vision" | "telemetry" | "rules" | "reasoning";

export type StewardSample = {
  t: number;
  date?: string | null;
  speed?: number | null;
  brake?: number | null;
  throttle?: number | null;
};

export type StewardSeries = {
  driver_number: number;
  samples: StewardSample[];
  location_points?: number;
};

export type StewardVerdict = {
  incident: string;
  rule_cited: string;
  telemetry_facts: string;
  verdict: string;
  penalty: string;
};

export type EvidenceStatus = "present" | "pending_phase2" | "insufficient";
export type SuccessProbability = "Low" | "Medium" | "High";

export type RequiredEvidenceItem = {
  id: string;
  label: string;
  rationale: string;
  status: EvidenceStatus;
  phase2_schema_ref: string;
};

export type RegulatoryCitation = {
  article_name: string;
  exact_quote: string;
  page_number: number;
  source_document: string;
};

export type ProtestDossier = {
  filing_type: "protest" | "right_of_review";
  filing_team: string;
  competitor_team: string;
  primary_claim: string;
  regulatory_violations: RegulatoryCitation[];
  available_evidence_summary: string;
  required_telemetry_evidence: RequiredEvidenceItem[];
  success_probability: SuccessProbability;
  legal_risk_notes: string;
  recommended_next_step: string;
  phase2_bridge: string;
};

/** Coerce legacy string citations into RegulatoryCitation objects for the dossier UI. */
export function normalizeProtestDossier(raw: ProtestDossier | null | undefined): ProtestDossier | null {
  if (!raw) return null;
  const violations = ((raw.regulatory_violations || []) as unknown[]).map((item) => {
    if (typeof item === "string") {
      return {
        article_name: item.slice(0, 120),
        exact_quote: item,
        page_number: 0,
        source_document: "unknown",
      } satisfies RegulatoryCitation;
    }
    const cite = (item || {}) as Partial<RegulatoryCitation>;
    return {
      article_name: cite.article_name || "",
      exact_quote: cite.exact_quote || "",
      page_number: Number(cite.page_number) || 0,
      source_document: cite.source_document || "",
    };
  });
  return { ...raw, regulatory_violations: violations };
}

export type LiveFeedContext = {
  session_type?: string;
  lap_number?: number;
  involved_driver_numbers?: number[];
  timing_note?: string;
};

export type StewardAnalyzeRequest = {
  clip_url?: string;
  year?: number;
  circuit?: string;
  meeting_key?: number;
  session_key?: number;
  incident_hint?: string;
  live_feed?: LiveFeedContext;
  filing_team?: string;
  filing_type?: "protest" | "right_of_review";
};

export type StewardAnalyzeResponse = {
  vision: Record<string, unknown>;
  telemetry_summary: string;
  telemetry_series: StewardSeries[];
  telemetry_degraded: boolean;
  session_key?: number | null;
  retrieved_rules: Array<{ id?: string; title?: string; text?: string; score?: number }>;
  verdict: StewardVerdict;
  protest_dossier: ProtestDossier;
  pipeline: Array<{ stage: StewardPipelineStage; status: string; detail?: string }>;
  assumptions: string[];
  errors: string[];
  disclaimer: string;
};

const DEFAULT_API_BASE = "http://127.0.0.1:8000";

function apiBase(): string {
  const base =
    typeof window === "undefined"
      ? process.env.API_INTERNAL_URL ?? process.env.NEXT_PUBLIC_API_URL ?? DEFAULT_API_BASE
      : process.env.NEXT_PUBLIC_API_URL ?? DEFAULT_API_BASE;
  return base.replace(/\/$/, "");
}

export async function analyzeStewardClip(body: StewardAnalyzeRequest): Promise<StewardAnalyzeResponse> {
  const response = await fetch(`${apiBase()}/api/steward/analyze_clip`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(180000),
  });
  if (!response.ok) {
    throw new Error(await response.text());
  }
  return response.json() as Promise<StewardAnalyzeResponse>;
}

export async function analyzeStewardUpload(input: {
  file: File;
  year?: number;
  circuit?: string;
  incidentHint?: string;
  clipUrl?: string;
  liveFeed?: LiveFeedContext;
  filingTeam?: string;
  filingType?: "protest" | "right_of_review";
}): Promise<StewardAnalyzeResponse> {
  const form = new FormData();
  form.append("file", input.file);
  if (input.year) form.append("year", String(input.year));
  if (input.circuit) form.append("circuit", input.circuit);
  if (input.incidentHint) form.append("incident_hint", input.incidentHint);
  if (input.clipUrl) form.append("clip_url", input.clipUrl);
  if (input.liveFeed) form.append("live_feed_json", JSON.stringify(input.liveFeed));
  if (input.filingTeam) form.append("filing_team", input.filingTeam);
  if (input.filingType) form.append("filing_type", input.filingType);
  const response = await fetch(`${apiBase()}/api/steward/analyze_clip/upload`, {
    method: "POST",
    body: form,
    signal: AbortSignal.timeout(180000),
  });
  if (!response.ok) {
    throw new Error(await response.text());
  }
  return response.json() as Promise<StewardAnalyzeResponse>;
}

export async function analyzeStewardClipStream(
  body: StewardAnalyzeRequest,
  onStage: (stage: StewardAnalyzeResponse["pipeline"][number]) => void,
): Promise<StewardAnalyzeResponse> {
  const response = await fetch(`${apiBase()}/api/steward/analyze_clip/stream`, {
    method: "POST",
    headers: { "Content-Type": "application/json", Accept: "text/event-stream" },
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(180000),
  });
  if (!response.ok || !response.body) {
    return analyzeStewardClip(body);
  }
  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  let result: StewardAnalyzeResponse | null = null;
  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });
    const chunks = buffer.split("\n\n");
    buffer = chunks.pop() ?? "";
    for (const chunk of chunks) {
      const eventLine = chunk.split("\n").find((line) => line.startsWith("event:"));
      const dataLine = chunk.split("\n").find((line) => line.startsWith("data:"));
      if (!dataLine) continue;
      const event = eventLine?.slice(6).trim();
      const data = JSON.parse(dataLine.slice(5).trim());
      if (event === "stage") onStage(data);
      if (event === "result") result = data as StewardAnalyzeResponse;
    }
  }
  if (!result) return analyzeStewardClip(body);
  return result;
}

File: backend/app/schemas/steward.py
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field, field_validator


class LiveFeedContext(BaseModel):
    """Timing / broadcast overlay facts available during a live session.

    This is NOT the steward verdict — only session context a real pitwall would
    already know (lap counter, car numbers, corner call from timing).
    """

    session_type: Optional[str] = None
    lap_number: Optional[int] = None
    involved_driver_numbers: List[int] = Field(default_factory=list)
    timing_note: str = ""

    @field_validator("involved_driver_numbers", mode="before")
    @classmethod
    def _coerce_driver_numbers(cls, value: Any) -> List[int]:
        if value is None:
            return []
        if not isinstance(value, list):
            return []
        out: List[int] = []
        for item in value:
            try:
                number = int(item)
            except (TypeError, ValueError):
                continue
            if 1 <= number <= 99:
                out.append(number)
        return list(dict.fromkeys(out))


class StewardAnalyzeRequest(BaseModel):
    clip_url: Optional[str] = None
    year: Optional[int] = None
    circuit: Optional[str] = None
    meeting_key: Optional[int] = None
    session_key: Optional[int] = None
    incident_hint: str = ""
    live_feed: Optional[LiveFeedContext] = None
    filing_team: str = "Mercedes-AMG Petronas Formula One Team"
    filing_type: Literal["protest", "right_of_review"] = "protest"


class StewardVerdict(BaseModel):
    """Legacy summary fields retained for compatibility with earlier steward UI."""

    incident: str = ""
    rule_cited: str = ""
    telemetry_facts: str = ""
    verdict: str = ""
    penalty: str = ""


EvidenceStatus = Literal["present", "pending_phase2", "insufficient"]
SuccessProbability = Literal["Low", "Medium", "High"]


class RequiredEvidenceItem(BaseModel):
    id: str = ""
    label: str = ""
    rationale: str = ""
    status: EvidenceStatus = "pending_phase2"
    phase2_schema_ref: str = ""


class RegulatoryCitation(BaseModel):
    """Verbatim regulation cite — quote must not be paraphrased by counsel."""

    article_name: str = ""
    exact_quote: str = ""
    page_number: int = 0
    source_document: str = ""

    @field_validator("page_number", mode="before")
    @classmethod
    def _coerce_page(cls, value: Any) -> int:
        try:
            return max(0, int(value))
        except (TypeError, ValueError):
            return 0


class ProtestDossier(BaseModel):
    """Phase 1 output: Team Principal Protest / Right of Review dossier."""

    filing_type: Literal["protest", "right_of_review"] = "protest"
    filing_team: str = "Mercedes-AMG Petronas Formula One Team"
    competitor_team: str = ""
    primary_claim: str = ""
    regulatory_violations: List[RegulatoryCitation] = Field(default_factory=list)
    available_evidence_summary: str = ""
    required_telemetry_evidence: List[RequiredEvidenceItem] = Field(default_factory=list)
    success_probability: SuccessProbability = "Low"
    legal_risk_notes: str = ""
    recommended_next_step: str = ""
    phase2_bridge: str = (
        "Ingest high-frequency micro-telemetry via POST /api/steward/phase2/telemetry "
        "to satisfy pending evidentiary items before lodging with the FIA."
    )

    @field_validator("regulatory_violations", mode="before")
    @classmethod
    def _coerce_violations(cls, value: Any) -> List[Any]:
        if value is None:
            return []
        if not isinstance(value, list):
            return []
        out: List[Any] = []
        for item in value:
            if isinstance(item, dict):
                out.append(item)
            elif isinstance(item, str) and item.strip():
                out.append(
                    {
                        "article_name": item.strip()[:120],
                        "exact_quote": item.strip(),
                        "page_number": 0,
                        "source_document": "teaching_corpus",
                    }
                )
        return out


class Phase2TelemetryChannel(BaseModel):
    """Future Phase 2 contract: one high-frequency channel for a driver."""

    driver_number: int
    channel: Literal[
        "steering_angle_deg",
        "brake_pressure_bar",
        "throttle_pct",
        "speed_kph",
        "lateral_g",
        "longitudinal_g",
        "gps_x",
        "gps_y",
    ]
    sample_hz: float = Field(ge=10, le=100, description="10Hz–100Hz micro-telemetry")
    timestamps_iso: List[str] = Field(default_factory=list)
    values: List[float] = Field(default_factory=list)
    window_label: str = ""  # e.g. "Turn 5: 100m board → apex"


class Phase2TelemetryIngestRequest(BaseModel):
    """API contract for Phase 2 high-frequency telemetry ingestion (future scope)."""

    protest_ref: str = ""
    session_key: Optional[int] = None
    lap_number: Optional[int] = None
    turn_label: str = ""
    channels: List[Phase2TelemetryChannel] = Field(default_factory=list)
    onboard_tcam_url: Optional[str] = None
    notes: str = ""


class StewardAnalyzeResponse(BaseModel):
    vision: Dict[str, Any] = Field(default_factory=dict)
    telemetry_summary: str = ""
    telemetry_series: List[Dict[str, Any]] = Field(default_factory=list)
    telemetry_degraded: bool = False
    session_key: Optional[int] = None
    retrieved_rules: List[Dict[str, Any]] = Field(default_factory=list)
    verdict: StewardVerdict = Field(default_factory=StewardVerdict)
    protest_dossier: ProtestDossier = Field(default_factory=ProtestDossier)
    pipeline: List[Dict[str, Any]] = Field(default_factory=list)
    assumptions: List[str] = Field(default_factory=list)
    errors: List[str] = Field(default_factory=list)
    disclaimer: str = (
        "Simulated Team Principal Protest Dossier for a portfolio demo. "
        "Teaching-corpus rules only — not an official FIA Protest or Steward Decision."
    )
